src/bson.c

```diff
@@ -763,23 +772,30 @@ PHONGO_API void zval_to_bson(zval *data, php_phongo_bson_flags_t flags, bson_t *
 
 	switch(Z_TYPE_P(data)) {
 		case IS_OBJECT:
-			if (instanceof_function(Z_OBJCE_P(data), php_phongo_persistable_ce TSRMLS_CC)) {
+			if (instanceof_function(Z_OBJCE_P(data), php_phongo_serializable_ce TSRMLS_CC)) {
+				zend_call_method_with_0_params(&data, NULL, NULL, BSON_SERIALIZE_FUNC_NAME, &obj_data);
 
-				if (flags & PHONGO_BSON_ADD_ODS) {
-					bson_append_binary(bson, PHONGO_ODM_FIELD_NAME, -1, 0x80, (const uint8_t *)Z_OBJCE_P(data)->name, strlen(Z_OBJCE_P(data)->name));
+				if (!obj_data) {
+					/* zend_call_method() failed */
+					return;
```

Comment:
Instead of returning here and after an exception below, we should break and then change the `if (ht_data && ht_data->nApplyCount > 1)` conditional later to also return if `ht_data` is `NULL`. Additionally, ensure we initialize `ht_data` to `NULL` at the top of this function.